### PR TITLE
refactor(semantic): make SemanticBuilder opaque

### PR DESCRIPTION
**Pull request by DonIsaac**
change visibility of building-related methods and properties of SemanticBuilder
from `pub` to `pub(crate)`.

### CODE REVIEW

**Comment by DonIsaac** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4851?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#4851** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4851?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#4850** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4850?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @DonIsaac and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by camc314 — APPROVED** [review]

**Comment by rzvxa** [comment]
This is lovely, It's so annoying having builder fields available from the outside world. But let's not merge it until we also get our other breaking changes ready.
Adding `AstNodeId` and reordering call expression fields are also breaking changes so it would be worth it to wait just so we can increase the minor version once for all of these.

**Review by rzvxa — APPROVED** [review]

**Comment by Boshen** [comment]
Let me test Rolldown before merging.

**Comment by overlookmotel** [comment]
Very much agree with this change. If Rolldown is relying on accessing properties/methods of `SemanticBuilder`, it shouldn't be, and we should provide alternative methods on `Semantic` for them to access the info they need.

**Pull request merged**
